src/query.jl

```diff
@@ -0,0 +1,257 @@
+""" ->
+function add_format{sym}(fmt::Type{DataFormat{sym}}, magic::Union(Tuple,AbstractVector,ByteString), extension)
+    m = canonicalize_magic(magic)
+    rng = searchsorted(magic_list, m, lt=magic_cmp)
+    isempty(rng) || error("magic bytes ", m, " are already registered")
+    haskey(sym2info, sym) && error("format ", fmt, " is already registered")
+    insert!(magic_list, first(rng), Pair(m, sym))  # m=>sym in 0.4
+    sym2info[sym] = (m, extension)
+    add_extension(extension, sym)
+    fmt
+end
+
+# For when "magic" is supplied as a function (see the HDF5 example in
+# registry.jl)
+function add_format{sym}(fmt::Type{DataFormat{sym}}, magic, extension)
+    haskey(sym2info, sym) && error("format ", fmt, " is already registered")
+    push!(magic_func, Pair(magic,sym))  # magic=>sym in 0.4
+    sym2info[sym] = (magic, extension)
+    add_extension(extension, sym)
+    fmt
+end
+
+@doc """
+`del_format(fmt::DataFormat)` deletes `fmt` from the format registry.
+""" ->
+function del_format{sym}(fmt::Type{DataFormat{sym}})
+    magic, extension = sym2info[sym]
+    rng = searchsorted(magic_list, magic, lt=magic_cmp)
+    @assert length(rng) == 1
+    deleteat!(magic_list, first(rng))
+    delete!(sym2info, sym)
+    del_extension(extension)
+    nothing
+end
+
+@doc """
+`info(fmt)` returns the magic bytes/extension information for
+`DataFormat` `fmt`.""" ->
+Base.info{sym}(::Type{DataFormat{sym}}) = sym2info[sym]
+
+canonicalize_magic{N}(m::NTuple{N,UInt8}) = m
+canonicalize_magic(m::AbstractVector{UInt8}) = tuple(m...)
+canonicalize_magic(m::ByteString) = canonicalize_magic(m.data)
+
+function add_extension(ext::ASCIIString, sym)
+    if haskey(ext2sym, ext)
+        v = ext2sym[ext]
+        if isa(v, Symbol)
+            ext2sym[ext] = Symbol[v, sym]
+        else
+            push!(ext2sym[ext], sym)
+        end
+        return
+    end
+    ext2sym[ext] = sym
+end
+function add_extension(ext::Union(Array,Tuple), sym)
+    for e in ext
+        add_extension(e, sym)
+    end
+end
+
+del_extension(ext::ASCIIString) = delete!(ext2sym, ext)
+function del_extension(ext::Union(Array,Tuple))
+    for e in ext
+        del_extension(e)
+    end
+end
+
+# magic_cmp results in magic_list being sorted in order of increasing
+# length(magic), then (among tuples with the same length) in
+# dictionary order. This ordering has the advantage that you can
+# incrementally read bytes from the stream without worrying that
+# you'll encounter an EOF yet still have potential matches later in
+# the list.
+function magic_cmp(p::Pair, t::Tuple)
+    pt = first(p)
+    lp, lt = length(pt), length(t)
+    lp < lt && return true
+    lp > lt && return false
+    pt < t
+end
+function magic_cmp(t::Tuple, p::Pair)
+    pt = first(p)
+    lp, lt = length(pt), length(t)
+    lt < lp && return true
+    lt > lp && return false
+    t < pt
+end
+
+
+abstract Formatted{F<:DataFormat}   # A specific file or stream
```

Comment:
Yes, and that's what I had originally. But somehow `AbstractFormat` sounds to me more like a format specification, whereas these are supposed to represent instances (e.g., particular files).